src/Runtime/python/PyExecutionSession.hpp

```diff
@@ -34,7 +34,8 @@ PYBIND11_MODULE(PyRuntime, m) {
       .def("entry_points", &onnx_mlir::PyExecutionSession::pyQueryEntryPoints)
       .def("set_entry_point", &onnx_mlir::PyExecutionSession::pySetEntryPoint,
           py::arg("name"))
-      .def("run", &onnx_mlir::PyExecutionSession::pyRun, py::arg("input"))
+      .def("run", &onnx_mlir::PyExecutionSession::pyRun, py::arg("input"),
+          py::arg("shape"), py::arg("stride"))
```

Comment:
@negiyas could you update all documents about `PyExecutionSession` and `PyOMCompileExecutionSession`? since their run methods now need two additional params. Or recommend users to switch to `PyOMExecutionSession` instead.
I see several places (and maybe more):
- https://github.com/onnx/onnx-mlir/blob/main/docs/UsingPyRuntime.md
- https://github.com/onnx/onnx-mlir/blob/main/docs/mnist_example/README.md

Reply:
@tungld I will update the documents and example codes after all APIs are finalized. Thanks.

Reply:
@tungld In the latest PR, no APIs for pythons are changed from the main branch, as you can see in the following comments.
Accordingly no documents need to be updated.

Reply:
This breaks compatibility with existing client applications as the interface to .run() now takes three inputs instead of one.

Do we need to pass in "shape" and "stride" separately? After the "input" is passed in to the interface, internally could we not get this info from input.shape and input.strides?

Reply:
I forget we can take in a list of arrays. Although I think the idea would still apply. Get the shape and stride info for each tensor in the run "input" list. That would also remove the need to check that each of the three lists (input, shape, stride) need to be the same shape and the user needing to create those additional lists when the numpy arrays already have that info. 

(also side nit, they should be called "inputs" "shapes" and "strides" since we expect lists containing 1 or more of those things.)

Reply:
@cjvolzka The reason the shape and stride are passed in separately is because the input tensor of a multi-dimensional array of objects (pybind11 does not support string type conversion yet so we have to use generic object type) must be "flattened" to one-dimension when calling the native C/C++ runtime. If we pass in a multi-dimensional input tensor and let pybind11 do the conversion, in C++ code we would get a vector of vector of vector ... and it's very cumbersome to "flatten" this into one-dimension, requiring copying each dimension manually. Instead, doing a conversion in python (using the ravel() method of the numpy array) typically involves just a "view" change without copying at all since the original multi-dimensional input tensor is in most cases already contiguous in memory (unless you did slice etc. operations to the tensor in which case ravel() will do a copy when flattening).

So the change of the input is not so much about how many additional parameters are introduced, but rather the requirement of always passing in an one-dimensional tensor to pybind11.

Reply:
@gongsu832 I just think whether pybind11 supports optional arguments so that we can pass a single input for all the cases rather than string. And, if input is string, we have an internal check to require the other two arguments.

Reply:
With the Python wrapper, it's a lot more flexible to offer the API to the end user, as long as internally we "flatten" the input tensor before invoking pybind11. So taking a single input numpy array (which has all the shape info available) certainly works. Although it's typically a good practice to make the APIs of different languages as similar as possible. In this case, the C/C++ and Java APIs both take an one-dimensional array and a shape array (C/C++ API needs to take additional parameters such as rank and data type since it's typeless). So Python API should in general follow that theme using basic Python array because not everybody may be using numpy array for whatever reason. The single input numpy array API can be offered as an "added sugar" on top of the "normal" API.

Reply:
It makes sense to unify to using one-dimensional array in pybind11, and the Python wrapper will do whatever it can to support other data types, such as ones in numpy, torch, etc. So what is missing is a convenient way to deploy the python wrapper, e.g. pip package.

Reply:
FYI: this is a guideline about creating a .so file to package python scripts: https://fransiska.github.io/2023/09/14/creating-so-files-from-python 

I wonder if it is good to generate a PyRuntime*.so to contain the python wrapper. Then users will need two .so files (PyRuntime*.so and PyRuntimeC*.so) to keep their python program unchanged. Or, it is even better if we can package both the python wrapper and PyRuntimeC*.so in a single .so file, so that there is no change at the user side.

Reply:
> @gongsu832 I just think whether pybind11 supports optional arguments so that we can pass a single input for all the cases rather than string. And, if input is string, we have an internal check to require the other two arguments.

If this is possible, this would be the way I'd like to see this go. We keep backwards compatibility while enabling string support.

> Although it's typically a good practice to make the APIs of different languages as similar as possible. In this case, the C/C++ and Java APIs both take an one-dimensional array and a shape array (C/C++ API needs to take additional parameters such as rank and data type since it's typeless). So Python API should in general follow that theme using basic Python array because not everybody may be using numpy array for whatever reason.

I agree with this in principle but not in this implementation. Since this is an already established API, we need to plan out with backwards incompatibility a primary concern and not an afterthought, even if the new way is more consistent.

For example, we keep the interface to .run() "as is" using numpy arryies and add a deprecated warning to use a new run_model() (name tbd). Then .run_model() has whatever new requirements we want (ie always be flattened, but see below for more). Then after a release or two, we can drop the "old" .run() since it is deprecated.

Second, if we are trying to keep the interfaces consistent between languages, the .run_model() should not take an array, numpy array or a list of those. We should have OMTensor and OMTensorList Python objects. I've actually had many clients using models with a single input tensor get tripped up that the .run() input needs to be a list of numpy arrays and not just the single numpy array. If we required creating and passing in an OMTensorList Python object, that would be a lot clearer.

Additionally (ie not required, but I'd love to see this) we should add an API to all the languages that return some sort of API version (each language interface having it's own version). Then client applications (or serving environments), that support "old" and "new" models created by onnx-mlir can query the model and know what APIs are available.

Reply:
There is no "run()" API in C/C++ and Java. It's a Python thing which unfortunately wasn't following the API theme of other languages. The current Python API is built directly on pybind11 and as a result said before couldn't support string data type. So that forces us to introduce a Python wrapper in front of pybind11. All user facing Python APIs should now be implemented in this Python wrapper rather than directly on pybind11. The Python API should implement some basic routines similar to those in C/C++ and Java. It can also add additional routines like "run()" for convenience and backward compatibility.

Reply:
> FYI: this is a guideline about creating a .so file to package python scripts: https://fransiska.github.io/2023/09/14/creating-so-files-from-python
> 
> I wonder if it is good to generate a PyRuntime*.so to contain the python wrapper. Then users will need two .so files (PyRuntime*.so and PyRuntimeC*.so) to keep their python program unchanged. Or, it is even better if we can package both the python wrapper and PyRuntimeC*.so in a single .so file, so that there is no change at the user side.

There is no need to do that. A Python package is just a directory that can have both .py and .so files. So for us, we can have a .py wrapper for all user facing APIs and a .so for pybind11 related code.

Reply:
Ok, after the onnx-mlir SIG tonight I finally understood that there's also a new PyRuntime.py that acts as a backwards compatible wrapper for the updated PyRuntimeC.cpython-*-linux-gnu.so files. 

For zDLC, at build time, we also pre-build the PyRuntimeC files for all versions of Python that we support and copy them into our deliverable container image. Since I missed that PyRuntime.py exists, it wasn't copied into our container which made me think we had to update our pythonapplications to use new PyRuntimeC directly.

Now that I know it exists, I'll experiment with the PyRuntime.py later this week. Sorry for all the confusion!